### DISCUSSION TITLE
Directus own tables migration & multiple DBs

### DISCUSSION THREAD

**Discussion by ghost** [directus/directus#3629] [Archive]
Hello, I was wondering :

1. I understand directus creates tables in the DBs to save configuration, but should it ? Is it planned to, for example, rely on another dedicated directus configuration DB OR redis instance to store configuration ? That way, the origin DB is not touched ?
2. Is it planned to have multiple DBs in the interface? For exemple, it would be interesting to have both Prod and Staging dbs, and using a menu, we can go from one to another, all in the same directus instance.

**Comment by philleepflorence** [comment]
1. Do you mean configuration or settings? Environment specific configuration usually goes in an environment or configuration file, such as .env or in D8, a config file. Project specific settings are stored in the DB as those are editable by the admin. 
2. I think that was solved with the multi-tenant feature in D8. But I don't think there is a plan to add that functionality  to D9, I would let the core team confirm that. For D9 it looks like the Directus App maps to a single DB.

**Reply by ghost** [reply]
Thanks @philleepflorence for responding.

1. The project specific settings. I find it abnormal to have them stored in the DB. At least it should be in a specific D9 DB, or dedicated redis instance.
2. I think ability to change DB in same interface using a menu is a very important feature.

**Comment by philleepflorence** [comment]
1. How is it abnormal if the settings are controlled by the Administrators of the App and API? Do you mean have 2 separate DBs for Directus and the project? Is the Directus prefix to all collections required to make Directus function not enough to separate the logic while keeping a single DB?
2. I will let the core team respond to this in detail, but as a contributor and advocate of Directus, the functionality of multi-tenancy added a lot of complexity for the dozens of projects I manage. I still use separate instances of the Directus App to manage separate projects and DBs, sometimes on separate zones and servers.

**Comment by benhaynes** [comment]
@zenati — I've read your responses, and appreciate the feedback. What @philleepflorence said is correct... and below are all the reasons why we removed that multitenancy from the system in this version:

1. In previous versions of Directus we had one instance manage multiple databases, similar to a RDBMS, but we found that the vast majority of our users only managed one DB.
2. There was no clean way to scope extensions to one database when in a multitenant per instance setup. It was getting very messy.
3. There was a significant amount of code needed to maintain that multitenancy.
4. Our new v9 permissions allow for multitenancy in a single database.
5. It's much easier to install Directus now (npm/Docker/k8s) so it's not a big deal to create new instances for each DB.
6. Our own Cloud system has multitenancy with one instance per database.

As for having system tables in the same database...

1. It's a lot of overhead for a "normal" user to setup two databases for one installation
2. Adding the "option" to use two DBs isn't off the table, but it more work than we (FOSS) have time to do right now
3. System tables (as of now) manage Users and Files, both of which _can_ be used relationally. We need them in the same database to do JOINs with non-system data.
4. It's not really that obstructive — and I don't think you'll find any other CMS that comes close to the decoupling and system data isolation that Directus has. We're very careful NOT to commingle any system data with user data... so keeping the tables in the same database isn't much of an issue. You're the first person I've heard be _super_ surprised by this in the 16 years since I created Directus.

I'm totally down to keep discussing this as an option, but please keep in mind that we have 3M+ users, and we have a [strict 80/20 rule](https://docs.directus.io/getting-started/contributing.html#feature-requests) about new features like this. In the end, this is open-source, so you can always fork and update the core codebase to suit your needs!

**Reply by bambino** [reply]
Hello,
First of all thanks for this great tool.
I agree with zenati, today setting up a databse can be done in seconds (containers ...)
so create another database dedicated for directus tables is not a problem.
you talked about "setup two databases for one installation", but i'm pretty sure most of the time Directus is setup on an already existing database, so no headache to add a second database when setting up Directus.
Often, databases are regularly dumped or replicated (to copy production data to non-production instances for example), if i want Directus on each database instance but with different parameters (users, rights, webhooks ...) so there it's a lot of overhead for a "normal" user.
Furthermore, some databases can be owned by integrated softwares packages and it could be great to add Directus on those without beeing invasive.
Maybe default usage can be to create specific tables on the aimed database but at least it could be great to allow creating tables somewhere else.

**Reply by benhaynes** [reply]
All good points. This may be considered in the future, but first we would need to allow mapping _custom_ user and file collections (to replace the system ones). Otherwise, with the users/files in the "system" database, we wouldn't be able to perform joins between them and the user's database content.

> System tables (as of now) manage Users and Files, both of which can be used relationally. We need them in the same database to do JOINs with non-system data.

**Comment by tnelson-doghouse** [comment]
For reference: https://github.com/directus/directus/discussions/2882

**Comment by dawidfreelancer** [comment]
That would be more clear to have Directus generated tables on separated database. 

1. It increases numbers of tables in database.
2. When i do fresh db setup on dev environment drop + run migrations all tables are gone so I have to recreate Directus tables.

For me ideally if all data that needs to be persistent was stored in file, so i can mount it to docker volume for persistence.

**Comment by diegogurpegui** [comment]
I'm in full support of having the ability to have all **directus tables in a separate database** if required. I found myself in the situation described by #2882 more than once.
